addons/payment/models/payment_acquirer.py

```diff
@@ -304,121 +279,94 @@ class PaymentTransaction(osv.Model):
-        if transaction.state not in ['cancel', 'error']:
-            if self.search(cr, uid, [('reference', '=', transaction.reference), ('id', '!=', transaction.id)], context=context, count=True):
-                return False
+    create_date = fields.Datetime(string='Creation Date', readonly=True)
+    date_validate = fields.Datetime(string='Validation Date')
+    acquirer_id = fields.Many2one('payment.acquirer', string='Acquirer', required=True)
+    payment_mode = fields.Selection([('server2server', 'Server To Server'), ('form', 'Form'), ('form_save', 'Form with credentials storage')], oldname='type', default="form", required=True)
+    state = fields.Selection(
+        [('draft', 'Draft'), ('pending', 'Pending'),
+         ('done', 'Done'), ('error', 'Error'),
+         ('cancel', 'Canceled')
+         ], 'Status', required=True, default="draft",
+        track_visibility='onchange', copy=False)
+    state_message = fields.Text(string='Message', help='Field used to store error and/or validation messages for information')
+    # payment
+    amount = fields.Float(required=True, digits=(16, 2), track_visibility='always', help='Transaction Amount')
+    fees = fields.Float(digits=(16, 2), track_visibility='always', help='Fees amount; set by the system because depends on the acquirer')
+    currency_id = fields.Many2one('res.currency', string='Currency', required=True)
+    reference = fields.Char(
+        default=lambda self: self.env['ir.sequence'].next_by_code('payment.transaction'),
+        required=True, help='Internal reference of the TX')
+    acquirer_reference = fields.Char(help='Reference of the TX as stored in the acquirer database')
+    # duplicate partner / transaction data to store the values at transaction time
+    partner_id = fields.Many2one('res.partner', string='Partner', track_visibility='onchange')
+    partner_name = fields.Char()
+    partner_lang = fields.Selection(_lang_get, string='Language', default="en_US")
+    partner_email = fields.Char(string='Email')
+    partner_zip = fields.Char(string='Zip')
+    partner_address = fields.Char(string='Address')
+    partner_city = fields.Char(string='City')
+    partner_country_id = fields.Many2one('res.country', string='Country', default=_default_partner_country_id, required=True)
+    partner_phone = fields.Char(string='Phone')
+    html_3ds = fields.Char(string='3D Secure HTML')
+
+    callback_eval = fields.Char(string='S2S Callback', help="""\
+        Will be safe_eval with `self` being the current transaction. i.e.:
+            self.env['my.model'].payment_validated(self)""", oldname="s2s_cb_eval")
+    payment_method_id = fields.Many2one('payment.method', string='Payment Method', domain="[('acquirer_id', '=', acquirer_id)]")
+
+    @api.constrains('reference', 'state')
+    def _check_reference(self):
+        if self.state not in ['cancel', 'error']:
+            if self.search([('reference', '=', self.reference), ('id', '!=', self.id)], count=True):
```

Comment:
dear @gpa-odoo instead of search_count so many ways like read_group also gives the nuber of recordset.